model/context.go

```diff
@@ -18,107 +18,264 @@
-		return nil, errors.New("Invalid type for fetching Page")
+		return nil, errors.New("Invalid type for fetching Context fields")
+	}
+
+	decoder := utility.ManualDecoder{}
+	ctxInp := decoder.MapStr(raw, "context")
+	userInp := decoder.Interface(ctxInp, "user")
+	err = decoder.Err
+	http, err := decodeHttp(ctxInp, err)
+	url, err := decodeUrl(ctxInp, err)
+	label, err := decodeLabel(ctxInp, err)
+	custom, err := decodeCustom(ctxInp, err)
+	page, err := decodePage(ctxInp, err)
+	user, err := metadata.DecodeUser(userInp, err)
+	if err != nil {
+		return nil, err
+	}
+	return &Context{
+		Http:   http,
+		Url:    url,
+		Label:  label,
+		Page:   page,
+		Custom: custom,
+		User:   user,
+	}, nil
+
+}
+
+func decodeUrl(raw common.MapStr, err error) (*Url, error) {
+	if err != nil {
+		return nil, err
+	}
+
+	decoder := utility.ManualDecoder{}
+	req := decoder.MapStr(raw, "request")
+	if req == nil {
+		return nil, decoder.Err
+	}
+
+	inpUrl := decoder.MapStr(req, "url")
+	url := Url{
+		Original: decoder.StringPtr(inpUrl, "raw"),
+		Full:     decoder.StringPtr(inpUrl, "full"),
+		Domain:   decoder.StringPtr(inpUrl, "hostname"),
+		Path:     decoder.StringPtr(inpUrl, "pathname"),
+		Query:    decoder.StringPtr(inpUrl, "search"),
+		Fragment: decoder.StringPtr(inpUrl, "hash"),
+	}
+	if scheme := decoder.StringPtr(inpUrl, "protocol"); scheme != nil {
+		trimmed := strings.TrimSuffix(*scheme, ":")
+		url.Scheme = &trimmed
+	}
+	if decoder.Err != nil {
+		return nil, decoder.Err
+	}
+
+	if url.Port = decoder.IntPtr(inpUrl, "port"); url.Port != nil {
+		return &url, nil
+	}
+
+	if portStr := decoder.StringPtr(inpUrl, "port"); portStr != nil {
+		if p, err := strconv.Atoi(*portStr); err == nil {
+			url.Port = &p
+		}
+	}
+
+	return &url, nil
+}
+
+func decodeHttp(raw common.MapStr, err error) (*Http, error) {
+	if err != nil {
+		return nil, err
+	}
+	var http *Http
+	decoder := utility.ManualDecoder{}
+	inpReq := decoder.MapStr(raw, "request")
+	if inpReq != nil {
+		headers := Headers(decoder.MapStr(inpReq, "headers"))
+		http = &Http{
+			Version: decoder.StringPtr(inpReq, "http_version"),
+			Request: &Req{
+				Method: strings.ToLower(decoder.String(inpReq, "method")),
+				Env:    decoder.Interface(inpReq, "env"),
+				Socket: &Socket{
+					RemoteAddress: decoder.StringPtr(inpReq, "remote_address", "socket"),
+					Encrypted:     decoder.BoolPtr(inpReq, "encrypted", "socket"),
+				},
+				Headers: &headers,
+				Body:    decoder.Interface(inpReq, "body"),
+				Cookies: decoder.Interface(inpReq, "cookies"),
+			},
+		}
+	}
+
+	inpResp := decoder.MapStr(raw, "response")
+	if inpResp != nil {
+		if http == nil {
+			http = &Http{}
+		}
+		headers := Headers(decoder.MapStr(inpResp, "headers"))
+		http.Response = &Resp{
+			Finished:    decoder.BoolPtr(inpResp, "finished"),
+			StatusCode:  decoder.IntPtr(inpResp, "status_code"),
+			HeadersSent: decoder.BoolPtr(inpResp, "headers_sent"),
+			Headers:     &headers,
+		}
+	}
+
+	if decoder.Err != nil {
```

Comment:
We're inconsistent about this but I'd probably have written `return http, decoder.Err` to eliminate the nil check and put a bit more burden on the caller to not use `http` in case of an error